src/funit/asserts/Assert_Real.tmpl

```diff
@@ -981,7 +1056,101 @@ contains
+      {expected.type} (kind={expected.kind}), intent(in) :: expected {expected.dims}
+      {actual.type} (kind={actual.kind}), intent(in) :: actual {actual.dims}
+      {rel_tolerance.type} (kind={rel_tolerance.kind}), intent(in) :: rel_tolerance
+      {tolerance.type} (kind={tolerance.kind}), intent(in) :: tolerance
+      character(*), optional, intent(in) :: message
+      type (SourceLocation), optional, intent(in) :: location
+
+      real(kind=kind(actual)) :: rel_t
+      real(kind=kind(actual)) :: t
+      real(kind=kind(actual)), allocatable :: e {actual.dims}
+
+      if (.not. conformable(shape(expected), shape(actual))) then
+         call fail_not_conformable(shape(expected), shape(actual), message=message, location=location)
+         return
+      end if
+
+      ! Trick to get e to have the right shape even if "expected" is a scalar.
+      e = 0*actual + expected
+
+      !if (any([e] == 0)) then
+      !   call fail_generic('Small denominator detected in AssertRelMinEqual.',message=message, location=location)
+      !   return
+      !end if
+
+      rel_t = real(rel_tolerance, kind(actual))
+      t = real(tolerance, kind(actual))
+
+      call assert_rel_min_equal(e, actual, rel_t, t, message=message, location=location)
+
+   end subroutine {name}
+   @end template
    
+   @template(assert_rel_min_equal,[actual])
+   subroutine {name}(expected, actual, rel_tolerance, tolerance, unused, message, location)
+      {actual.type} (kind={actual.kind}), intent(in) :: expected {actual.dims}
+      {actual.type} (kind={actual.kind}), intent(in) :: actual {actual.dims}
+      {actual.type} (kind={actual.kind}), intent(in) :: rel_tolerance
+      {actual.type} (kind={actual.kind}), intent(in) :: tolerance
+      class (KeywordEnforcer), optional, intent(in) :: unused
+      character(*), optional, intent(in) :: message
+      type (SourceLocation), optional, intent(in) :: location
+
+      real(kind=kind(actual)) :: e, a, rd, calc_eps
+
+#if {actual.rank} != 0
+      integer, allocatable :: i(:)
+#endif
+      character(len=MAX_LEN_REAL_AS_STRING) :: expected_str
+      character(len=MAX_LEN_REAL_AS_STRING) :: actual_str
+      character(len=3*MAX_LEN_REAL_AS_STRING) :: diff_str
+
+      _UNUSED_DUMMY(unused)
+
+#if {actual.rank} == 0
+      ! scalar
+      if (.not. (abs(actual - expected) <= max(rel_tolerance*abs(expected), tolerance))) then
+         e = expected
+         a = actual
+      else
+         return
+      end if
+         
+#else
+      if (.not. all(abs(actual - expected) <= max(rel_tolerance*abs(expected), tolerance))) then
+         ! index of first difference is
+         i = maxloc(merge(1,0, .not. abs(actual-expected) <= max(rel_tolerance*abs(expected), tolerance)))
+         e = expected({actual.multi_index})
+         a = actual({actual.multi_index})
+      else
+         return
+      end if
+#endif
+      calc_eps = max(rel_tolerance*abs(e), tolerance)
+      if ( e > 0) then
```

Comment:
OK - with this line, I finally see what you are trying to achieve.       But what do you think of instead adding a check in the existing procedure that generates a fail message if the denominator is zero?

Or do you really have a use case where you'd want to use a relatively equal on a value that is potentially exactly zero?     All of the cases I can readily imagine, I'd want diagnosed.

Reply:
> Or do you really have a use case where you'd want to use a relatively equal on a value that is potentially exactly zero? All of the cases I can readily imagine, I'd want diagnosed.

Unfortunately yes. Because this is a refactor of regression tests, I haven't been thinking terribly hard about the underlying data. The old test does a relative assert over an array of data, and the first iteration in the loop compares 0.0 to 0.0.

I will spend some time this week re-evaluating these tests to see if I can get around needing this relative assert that defaults to regular tolerance. 

As I have thought of this more. I would like for this to make sense, but needs more logic to make it more clear what is going on. 

I like the way `approx` is described in [pytest](https://github.com/pytest-dev/pytest/blob/33c6ad5bf76231f1a3ba2b75b05ea2cd728f9919/src/_pytest/python_api.py)

Reply:
Thank you for the link.   Having a reference to an existing implementation is very useful.

Note that the style in pytest would be friendlier in fHamcrest.     You might want to consider an extension in that layer as well.

The part that still makes me cringe a little is the somewhat arbitrary defaults in the pytest version.    In Fortran, I'd replace 1.e-6 with `epsilon(e)` or maybe `2*epsilon(e)`, but the default for the absolute tolerance is harder to justify from a generic numerical vantage point.   Possibly it should be some multiple of `tiny(e)` or some small multiple of that.   I'm glad your implementation side-steps such issues by requiring that they be specified.